From a296273434c72f288ad140339f3584604e8033ef Mon Sep 17 00:00:00 2001
From: Dave Taht <dave.taht@bufferbloat.net>
Date: Thu, 29 Mar 2012 20:34:42 -0700
Subject: [PATCH] iw: chain signal support

---
 nl80211.h |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/nl80211.h b/nl80211.h
index e474f6e..12f3412 100644
--- a/nl80211.h
+++ b/nl80211.h
@@ -1708,6 +1708,8 @@ enum nl80211_sta_info {
 	NL80211_STA_INFO_CONNECTED_TIME,
 	NL80211_STA_INFO_STA_FLAGS,
 	NL80211_STA_INFO_BEACON_LOSS,
+	NL80211_STA_INFO_CHAIN_SIGNAL,
+	NL80211_STA_INFO_CHAIN_SIGNAL_AVG,
 
 	/* keep last */
 	__NL80211_STA_INFO_AFTER_LAST,
-- 
1.7.1

